### PR TITLE
Update self-push workflow

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

Update self-push CI workflow file:

- `tests_fetcher.py` is updated to output a json file, containing a dictionary mapping test categories to the identified test files (which is used by the updated push CI below)
- Reorganize the tests into models (e.g. `models/bert`, `models/gpt2`, etc.) and modeling categories (`pipeline`, `tokenization`), same as in scheduled CI
- `notification_service.py` and `self-scheduled.yml` are updated to use `[single/multi]-gpu` as artifact name prefixes (i.e. no more `-docker` at the end): with this minimal change, `notification_service.py` could be reused

Some workflow runs:
- [push CI](https://github.com/huggingface/transformers/actions/runs/2306332297)
- [scheduled CI](https://github.com/huggingface/transformers/actions/runs/2306421236)

Some tests failed intentionally (to verify their reports). The reports could be found on `transformers-ci-feedback-tests` channel.

### TODO:
  - create new report channel and add the channel ID to the workflow file

**I added some reviews that contain some of my questions.**

@sgugger Maybe you could have a look for the changes in `test_fetcher.py`?
@stas00 Maybe for the changes regarding DeepSpeed and multi-gpu configurations?

### CODE REVIEW

File: .github/workflows/self-push.yml [hunk 1]
```diff
@@ -20,37 +20,34 @@ env:
   OMP_NUM_THREADS: 8
   MKL_NUM_THREADS: 8
   PYTEST_TIMEOUT: 60
+  SIGOPT_API_TOKEN: ${{ secrets.SIGOPT_API_TOKEN }}
+  TF_FORCE_GPU_ALLOW_GROWTH: true
+  # TODO: Should we enable this?
+  # RUN_PT_TF_CROSS_TESTS: 1
```
Comment (ydshieh) [inline .github/workflows/self-push.yml:26]: Current push CI don't run PT/TF cross tests. Guess it was because we use framework specific docker images.

Now we use `transformers-all-latest-gpu` (as scheduled CI), do you think it's good to enable it?

File: .github/workflows/self-push.yml [hunk 2]
```diff
@@ -20,37 +20,34 @@ env:
...
-      - name: Are GPUs recognized by our DL frameworks
+      - name: Cleanup
         run: |
-          utils/print_env_pt.py
+          rm -rf tests/__pycache__
+          rm -rf tests/models/__pycache__
+          rm -rf reports
 
       - name: Fetch the tests to run
+        # TODO: add `git-python` in the docker images
```
Comment (ydshieh) [inline .github/workflows/self-push.yml:48]: `pip install --upgrade git-python` is required so far.

File: .github/workflows/self-push.yml [hunk 3]
```diff
@@ -59,441 +56,221 @@ jobs:
...
-          name: test_fetched
-          path: test_preparation.txt
-
-      - name: Run all non-slow tests on GPU
-        env:
-          MKL_SERVICE_FORCE_INTEL: 1
+          echo "${{ matrix.folders }}"
+          echo "${{ fromJson(needs.setup.outputs.test_map)[matrix.folders] }}"
+          matrix_folders=${{ matrix.folders }}
+          echo "$matrix_folders"
```
Comment (ydshieh) [inline .github/workflows/self-push.yml:111]: No need to echo $matrix_folders --> I can remove it before merge.
(Already removed in scheduled CI)

File: .github/workflows/self-push.yml [hunk 4]
```diff
@@ -59,441 +56,221 @@ jobs:
...
+      matrix:
+        folders: ${{ fromJson(needs.setup.outputs.matrix) }}
+        machines: [multi-gpu]
+    runs-on: [self-hosted, docker-gpu, '${{ matrix.machines }}']
     container:
-      image: nvcr.io/nvidia/pytorch:21.03-py3
-      options: --gpus 0 --shm-size "16gb" --ipc host -v /mnt/cache/.cache/huggingface:/mnt/cache/
+      image: huggingface/transformers-all-latest-gpu
+      # TODO: Should it be `all` or `0`. Same question for places in other workflow files.
+      options: --gpus all --shm-size "16gb" --ipc host -v /mnt/cache/.cache/huggingface:/mnt/cache/
```
Comment (ydshieh) [inline .github/workflows/self-push.yml:151]: In push CI, currently we have `--gpus all`
In scheduled CI, `run_tests_multi_gpu` has `options: --gpus 0`.

Is that a mistake, and we should instead using `--gpus all` for multi-gpu testing both in push/scheduled CI?

File: .github/workflows/self-push.yml [hunk 5]
```diff
@@ -59,441 +56,221 @@ jobs:
...
-
       - name: NVIDIA-SMI
         run: |
           nvidia-smi
 
-      - name: Install dependencies
-        run: |
-          apt -y update && apt install -y libaio-dev
-          pip install --upgrade pip
-          pip install .[deepspeed-testing]
```
Comment (stas00) [inline .github/workflows/self-push.yml:387]: where the deps are now installed? I don't see them in the new version

**Review by stas00 — COMMENTED** [review]
I find it difficult to make sense of some of the changes, due to diff being hard to follow, so I'm totally fine with you going ahead with the deepspeed side of things and sort it out if something breaks afterwards. I test on my machine as well, so if it's broken for a day or two on CI it's no big deal.

**Comment by ydshieh** [comment]
> I find it difficult to make sense of some of the changes, due to diff being hard to follow

OK @stas00 , the big diff is probably because I removed unused blocks. No real big change - I just copied from scheduled CI.
See below if you would like to have a quick look. 

My only questions are

  - why we use `options: --gpus 0` previously in `run_tests_torch_cuda_extensions_multi_gpu`
  - could we use `options: --gpus all` for single gpu case as well as multi gpu?

prev.
```
image: nvcr.io/nvidia/pytorch:21.03-py3
options: --gpus 0
```
now.
```
huggingface/transformers-pytorch-deepspeed-latest-gpu
options: --gpus all
```

and

prev.
```
      - name: Install dependencies
        run: |
          apt -y update && apt install -y libaio-dev
          pip install --upgrade pip
          pip install .[deepspeed-testing]
```
now
```
      - name: Re-compile DeepSpeed
        working-directory: /workspace
        run: |
          pip install deepspeed # installs the deps correctly
          rm -rf DeepSpeed
          git clone https://github.com/microsoft/DeepSpeed && cd DeepSpeed && rm -rf build
          DS_BUILD_CPU_ADAM=1 DS_BUILD_AIO=1 DS_BUILD_UTILS=1 python3 -m pip install -e . --global-option="build_ext" --global-option="-j8" --no-cache -v --disable-pip-version-check
```

File: .github/workflows/self-push.yml (hunk 5 above)
Comment (ydshieh) [reply .github/workflows/self-push.yml:387]: Now I change to the docker image
```
transformers-pytorch-deepspeed-latest-gpu
```
and inside the Dockerfile, I could see
```
RUN python3 -m pip install --no-cache-dir -e ./transformers[deepspeed-testing]
```

This is what we have in scheduled CI, but there is a step named `name: Re-compile DeepSpeed` too.

**Comment by stas00** [comment]
thank you for highlighting the changes @ydshieh - that's super helpful.

1. So for gpus:

- multi-gpu should be all
- single-gpu should be 0

so the original is correct.

2. For dependencies the original is correct.

Have a look at what it signifies:
```
extras["deepspeed-testing"] = extras["deepspeed"] + extras["testing"] + extras["optuna"]
```

so the change is missing important dependencies install.

and the new instructions aren't correct.

We only want the bleed edge (your now) install only for nightly build. self-push should use the released `deepspeed` version.

i.e. no change from the original in either case.

File: .github/workflows/self-push.yml (hunk 5 above)
Comment (stas00) [reply .github/workflows/self-push.yml:387]: if it's in the docker, then we don't need it in the workflow file.

and no need to re-compile deepspeed for self-push.

**Comment by ydshieh** [comment]
> so looking at the diff the original seems to be correct. perhaps not everywhere?

The current main branch has a job `run_tests_torch_cuda_extensions_multi_gpu` in `self-push.yml` which has `--gpus 0`.
This PR uses something like 
```
matrix:
    machines: [single-gpu, multi-gpu]
runs-on: [self-hosted, docker-gpu, '${{ matrix.machines }}']
```
and so far it uses `--gpus all`. I will need to set `--gpus` correctly depending on `'${{ matrix.machines }}'`. I am a bit surprised that `--gpus all` couldn't be used for only 1 GPU.

(furthermore, some places in `self-scheduled.yml`) have to be fixed too)

> 2. For dependencies the original is correct.
> self-push should use the released `deepspeed` version, that `pip install .[deepspeed-testing]` 

I will change back to the original version for this part, thank you.

**Comment by ydshieh** [comment]
Convert to draft while fixing a bug.

**Comment by stas00** [comment]
As long as the tests are run with `CUDA_VISIBLE_DEVICES=0` for `run_tests_single_gpu` jobs it indeed doesn't matter if more than 1 gpu is available.

But it's critical we ensure that it is set correctly, otherwise tests requiring a single gpu will get skipped. 

Thank you for fixing where the setting are incorrect, @ydshieh!

File: .github/workflows/self-push.yml [hunk 6]
```diff
@@ -20,37 +20,34 @@ env:
   OMP_NUM_THREADS: 8
   MKL_NUM_THREADS: 8
   PYTEST_TIMEOUT: 60
+  SIGOPT_API_TOKEN: ${{ secrets.SIGOPT_API_TOKEN }}
```
[the remark this answers is not in the document]
Comment (LysandreJik) [reply .github/workflows/self-push.yml:23]: This is for `SIGOPT`, which only runs with slow tests. This can be removed!

File: .github/workflows/self-push.yml (hunk 1 above)
Comment (LysandreJik) [reply .github/workflows/self-push.yml:26]: Yes, would be good indeed

File: .github/workflows/self-push.yml (hunk 3 above)
Comment (LysandreJik) [reply .github/workflows/self-push.yml:111]: If it helps with debugging, it'll likely help again. No problem for me to leave logs if you think they might be useful when these jobs will crash.

File: .github/workflows/self-push.yml [hunk 7]
```diff
@@ -444,56 +279,46 @@ jobs:
...
-
+          CI_SLACK_CHANNEL_ID_DAILY: ${{ secrets.CI_SLACK_CHANNEL_ID_DAILY }}
+          CI_SLACK_CHANNEL_DUMMY_TESTS: ${{ secrets.CI_SLACK_CHANNEL_DUMMY_TESTS }}
+          CI_EVENT: push
+        # We pass `needs.setup.outputs.matrix` as the argument. A processing in `notification_service.py` to change
+        # `models/bert` to `models_bert` is required, as the artifact names use `_` instead of `/`.
         run: |
           pip install slack_sdk
-          python utils/notification_service_deprecated.py push
+          python utils/notification_service.py "${{ needs.setup.outputs.matrix }}"
```
Comment (LysandreJik) [inline .github/workflows/self-push.yml:324]: Love that!

File: .github/workflows/self-scheduled.yml [hunk 8]
```diff
@@ -310,6 +353,7 @@ jobs:
           CI_SLACK_CHANNEL_ID: ${{ secrets.CI_SLACK_CHANNEL_ID }}
           CI_SLACK_CHANNEL_ID_DAILY: ${{ secrets.CI_SLACK_CHANNEL_ID_DAILY }}
           CI_SLACK_CHANNEL_DUMMY_TESTS: ${{ secrets.CI_SLACK_CHANNEL_DUMMY_TESTS }}
+          CI_EVENT: scheduled
```
Comment (LysandreJik) [inline .github/workflows/self-scheduled.yml:356]: Smart!

**Review by LysandreJik — APPROVED** [review]
This looks great, thanks for working on it @ydshieh!

The `notification_service_deprecated.py` can likely be removed now.

Before merging it, could you do a test run when modifying the `setup.py` to ensure that all tests are run correctly? Thank you!

**Comment by ydshieh** [comment]
> Before merging it, could you do a test run when modifying the `setup.py` to ensure that all tests are run correctly? Thank you!

I had to fix a bug (i.e. when the test list is `tests`, i.e. when `setup.py` is changed).
A full test workflow run is [here](https://github.com/huggingface/transformers/actions/runs/2318670278).
After looking some failures, I am convinced that this PR is ready to be merged (the failures are the same as in scheduled CI runs).

Thank you for the reviews!
